frontend/src/services/UploadService.js

```diff
@@ -40,16 +40,36 @@ class UploadService {
         })
     }
 
-    uploadFile(shareId, file, partUrls, onProgress) {
+    uploadFile(shareId, file, partUrls, onProgress, currSpeed) {
         const numParts = partUrls.length;
         const partSize = Math.ceil(file.size / numParts);
 
         const results = [];
         const progress = Array(numParts).fill(0);
 
+        const prevSpeedsSize = 10
+        let prevSpeeds = Array(prevSpeedsSize).fill(0);
+        let previous = 0;
+
+        let first = true;
+        let i = 0;
         const progressUpdateTimer = setInterval(() => {
-            const current = progress.reduce((previousValue, currentValue) => previousValue + currentValue);
+
+            const current = progress.reduce((prev, curr) => prev + curr);
+            prevSpeeds[i] = current - previous;
+            previous = current;
+
+            const avSpeed = first
+                ? prevSpeeds.slice(0,i+1).reduce((prev, curr) => prev + curr) / (i+1)
+                : prevSpeeds.reduce((prev, curr) => prev + curr) / prevSpeedsSize;
+
+            currSpeed(avSpeed * 5);
             onProgress(current/file.size);
+
+            if (++i === prevSpeedsSize){
+                first = false;
+                i = 0;
+            }
```

Comment:
Optional recommendation: Save only the current progress in the array, calculate speed on each update as ` i+1 % prevSpeedsSize`, with additional modifications for the start and the correct conversion to a per second amount.